src/Storages/Hive/HiveCommon.h

```diff
@@ -0,0 +1,139 @@
+            std::shared_ptr<Apache::Hadoop::Hive::Table> table_,
+            const std::map<String, PartitionInfo> & partition_infos_,
+            ContextPtr context_)
+            : WithContext(context_)
+            , db_name(db_name_)
+            , table_name(table_name_)
+            , table(table_)
+            , partition_infos(partition_infos_)
+            , empty_partition_keys(table->partitionKeys.empty())
+        {
+        }
+
+
+        std::map<String, PartitionInfo> & getPartitionInfos()
+        {
+            std::lock_guard lock{mutex};
+            return partition_infos;
+        }
+
+        std::shared_ptr<Apache::Hadoop::Hive::Table> getTable() const
+        {
+            std::lock_guard lock{mutex};
+            return table;
+        }
+
+        std::vector<Apache::Hadoop::Hive::Partition> getPartitions() const;
+
+        std::vector<FileInfo> getFilesByLocation(const HDFSFSPtr & fs, const String & location);
+
+    private:
+        String db_name;
+        String table_name;
+
+        mutable std::mutex mutex;
+        std::shared_ptr<Apache::Hadoop::Hive::Table> table;
+        std::map<String, PartitionInfo> partition_infos;
+        const bool empty_partition_keys;
+
+        Poco::Logger * log = &Poco::Logger::get("HiveMetastoreClient");
+    };
+
+    explicit HiveMetastoreClient(std::shared_ptr<Apache::Hadoop::Hive::ThriftHiveMetastoreClient> client_, ContextPtr context_)
+        : WithContext(context_), client(client_), table_metadata_cache(1000)
+    {
+    }
+
+    std::shared_ptr<HiveTableMetadata> getTableMetadata(const String & db_name, const String & table_name);
+    void clearTableMetadata(const String & db_name, const String & table_name);
+    void setClient(std::shared_ptr<Apache::Hadoop::Hive::ThriftHiveMetastoreClient> client_);
+    bool isExpired() const { return expired; }
+    void setExpired() { expired = true; }
+    void clearExpired() { expired = false; }
+
+private:
+    static String getCacheKey(const String & db_name, const String & table_name)  { return db_name + "." + table_name; }
+
+    bool shouldUpdateTableMetadata(
+        const String & db_name, const String & table_name, const std::vector<Apache::Hadoop::Hive::Partition> & partitions);
+
+    std::shared_ptr<Apache::Hadoop::Hive::ThriftHiveMetastoreClient> client;
+    LRUCache<String, HiveTableMetadata> table_metadata_cache;
+    mutable std::mutex mutex;
+    std::atomic<bool> expired{false};
+
+    Poco::Logger * log = &Poco::Logger::get("HiveMetastoreClient");
+};
+
+using HiveMetastoreClientPtr = std::shared_ptr<HiveMetastoreClient>;
+
+class HiveMetastoreClientFactory final : private boost::noncopyable
+{
+public:
+    static HiveMetastoreClientFactory & instance();
+
+    HiveMetastoreClientPtr getOrCreate(const String & name, ContextPtr context);
+
+private:
+    std::mutex mutex;
+    std::map<String, HiveMetastoreClientPtr> clients;
+
+    const int conn_timeout_ms = 60000;
```

Comment:
btw may be connect_timeout as 60 seconds too much?

Reply:
I copied from HDFS, how much do you think is better ? 
```
    hdfsBuilderConfSetStr(builder.get(), "input.read.timeout", "60000"); // 1 min
    hdfsBuilderConfSetStr(builder.get(), "input.write.timeout", "60000"); // 1 min
    hdfsBuilderConfSetStr(builder.get(), "input.connect.timeout", "60000"); // 1 min
```

Reply:
I think 10 sec is more sane. Though it is not very important. Usually it is better to make it configurable. But now I don't think it is really needed (not in demand).

Reply:
done